src/apps/filtering.js

```diff
@@ -1,10 +1,124 @@
 
-export const filterApps = (apps: App[], _options: FilterOptions): App[] => apps;
+const missingMarketcapTickers =
+  "In order to sort by marketcap you need to provide the marketcap tickers.";
+
+const missingCurrencyId =
+  "Although we are already filtering the apps to currency apps, this is needed for flow.";
+
+const searchFilter = (query?: string) => ({ name, currencyId }) => {
+  if (!query) return true;
+  const currency = currencyId ? getCryptoCurrencyById(currencyId) : null;
+  const terms = `${name} ${
+    currency ? `${currency.name} ${currency.ticker}` : ""
+  }`;
+  return terms.toLowerCase().includes(query.toLowerCase().trim());
+};
+
+const typeFilter = (type = "all", installedApps?: InstalledItem[]) => app => {
+  switch (type) {
+    case "installed":
+      invariant(installedApps, missingInstalledApps);
+      return installedApps.find(a => a.name === app.name);
+    case "not_installed":
+      invariant(installedApps, missingInstalledApps);
+      return !installedApps.find(a => a.name === app.name);
+    case "updatable":
+      invariant(installedApps, missingInstalledApps);
+      return installedApps.find(a => a.name === app.name && !a.updated);
+    case "supported":
+      return (
+        app.currencyId &&
+        isCurrencySupported(getCryptoCurrencyById(app.currencyId))
+      );
+    default:
+      return true;
+  }
+};
+
+export const sortApps = (apps: App[], _options: SortOptions): App[] => {
+  const { type, order = "asc", marketcapTickers } = _options;
+  const asc = order === "asc";
+  let sortedApps = [...apps];
+
+  switch (type) {
+    case "marketcap":
+      invariant(marketcapTickers, missingMarketcapTickers);
+      const [currencyApps, otherApps] = sortedApps.reduce(
+        ([currencyApps, otherApps], app) => {
+          return app.currencyId
+            ? [[...currencyApps, app], otherApps]
+            : [currencyApps, [...otherApps, app]];
+        },
+        [[], []]
+      );
+
+      currencyApps.sort((a1, a2) => {
+        const [app1, app2] = asc ? [a2, a1] : [a1, a2];
+
+        invariant(app1.currencyId, missingCurrencyId);
+        const index1 = marketcapTickers.indexOf(
+          getCryptoCurrencyById(app1.currencyId).ticker
+        );
+
+        invariant(app2.currencyId, missingCurrencyId);
+        const index2 = marketcapTickers.indexOf(
+          getCryptoCurrencyById(app2.currencyId).ticker
+        );
+
+        if (index1 === index2) return 0;
+        if (index1 === -1) return 1;
+        if (index2 === -1) return -1;
+
+        return index1 - index2;
+      });
+
+      otherApps.sort((a1, a2) => {
+        const [app1, app2] = asc ? [a1, a2] : [a2, a1];
+        return app1.name.localeCompare(app2.name);
+      });
+
+      return [...currencyApps, ...otherApps];
+    default:
+      sortedApps.sort((a1, a2) => {
+        const [app1, app2] = asc ? [a1, a2] : [a2, a1];
+        return app1.name.localeCompare(app2.name);
+      });
+  }
+
+  return sortedApps;
```

Comment:
I think the code above here is way too complex for what it's doing. the O() complexity is not that great here. it's better to build up an array for performance.